### PR TITLE
[BEAM-8575] Test flatten a single PC and test flatten a flattened PC

### PR DESCRIPTION
**Pull request by liumomo315**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by liumomo315** [comment]
R: @robertwb

Hi Robert, two simple tests for flatten, also tag more tests with validates runner attr to allow Jenkins to pick them up and run them on Datafow service.

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 1]
```diff
@@ -536,12 +538,23 @@ def test_flatten_no_pcollections(self):
...
+    pipeline.run()
+
+  @attr('ValidatesRunner')
   def test_flatten_same_pcollections(self):
     pipeline = TestPipeline()
     pc = pipeline | beam.Create(['a', 'b'])
     assert_that((pc, pc, pc) | beam.Flatten(), equal_to(['a', 'b'] * 3))
     pipeline.run()
 
+  @attr('ValidatesRunner')
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: This one doesn't need to be ValidatesRunner as it's an SDK feature.

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 2]
```diff
@@ -550,11 +563,23 @@ def test_flatten_pcollections_in_iterable(self):
     assert_that(result, equal_to([0, 1, 2, 3, 4, 5, 6, 7]))
     pipeline.run()
 
+  @attr('ValidatesRunner')
+  def test_flatten_a_flattened_pcollection(self):
+    pipeline = TestPipeline()
+    pcoll_1 = pipeline | 'Start 1' >> beam.Create([0, 1, 2, 3])
+    pcoll_2 = pipeline | 'Start 2' >> beam.Create([4, 5, 6, 7])
+    result = ((pcoll_1, pcoll_2) | 'Flatten' >> beam.Flatten(),
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: This is a bit hard to read. 

It also would probably be good to have a third collection in the mix, e.g. 

```
pcoll_12 = (pcoll_1, pcoll_2) | beam.Flatten()
pcoll_123 = (pcoll_12, pcoll_3) | 'FlattenAgain' >> beam.Flatten()
```

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 3]
```diff
@@ -550,11 +563,23 @@ def test_flatten_pcollections_in_iterable(self):
...
+              ) |'Flatten again' >> beam.Flatten()
+    assert_that(result, equal_to([x for x in range(8)]))
+    pipeline.run()
+
+  @attr('ValidatesRunner')
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: This raises an error on construction, should not be ValidatesRunner.

File: sdks/python/apache_beam/transforms/ptransform_test.py [hunk 4]
```diff
@@ -550,11 +563,23 @@ def test_flatten_pcollections_in_iterable(self):
...
   def test_flatten_input_type_must_be_iterable(self):
     # Inputs to flatten *must* be an iterable.
     with self.assertRaises(ValueError):
       4 | beam.Flatten()
 
+  @attr('ValidatesRunner')
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/ptransform_test.py]: Same.

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 1 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/ptransform_test.py]: Reverted.

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 2 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/ptransform_test.py]: Refactored as suggested:)

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 3 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/ptransform_test.py]: I see. Reverted this one too.

File: sdks/python/apache_beam/transforms/ptransform_test.py (hunk 4 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/ptransform_test.py]: Done.

**Comment by liumomo315** [comment]
Thanks, Robert! PTAL:)

**Comment by robertwb** [comment]
Flaky ERROR: test_read_from_text_with_file_name_file_pattern (apache_beam.io.textio_test.TextSourceTest)

**Pull request merged**